### ISSUE TITLE
Gulp: SASS linting

### ISSUE THREAD

**Issue by gewoonderrik** [trendwerk/trendpress#389]
We should re-add SASS linting and fix as many errors as possible. We also need to add the config files and output xml to our gitignores.

**Comment by sboerrigter** [comment]
In Gulp we could use something like this:

```
/**
 * Lint Sass
 */
gulp.task('scss-lint', function() {
  gulp.src('assets/styles/**/*.scss')
  .pipe(scsslint({
    'config': 'lint.yml',
  }))
});
```

And change the Sass watch task to:

```
gulp.watch('assets/styles/**/*.scss',['sass', 'scss-lint']);
```

**Comment by sboerrigter** [comment]
Feedback by @haroldangenent: The linter settings should probably not be in the root of the theme

**Comment by gewoonderrik** [comment]
We should wait until the Gulp branche has been merged before we start working on this.

**Comment by gewoonderrik** [comment]
I'm working on validating our `.scss` with a Grunt Lint plug-in (which I will not commit of course). I'm currently running into the following issues:

* **Vendor prefixes**: The Sass guideline states that you should never have to write vendor prefixes yourself. Best way to take care of this is by using an autoprefixer tool. We should implement one into Grunt or Gulp. The tool also takes care of checking the actual vendor-prefixes used are still needed which is a nice feature.
* **SelectorFormat** (see https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#selectorformat): WordPress doesn't use hyphenated lowercase in classes. This will always result into errors. There's not much we can do about this. Maybe we should disable this lint rule
* **QualifyingElement** (see https://github.com/CSSLint/csslint/wiki/Disallow-unqualified-attribute-selectors): This basically means that all the declarations in resets etc. are invalid. We should have a look at these files to see of all the things declared here are still necessary. This however won't fix all the problems so we should probably exclude base files from being linted. 

How does everyone feel about this?

**Comment by sboerrigter** [comment]
**Vendor prefixes**
I like the autoprefixer tool. Can we implement it? Maybe it should be another issue.

**SelectorFormat**
I am not sure about this. Do we need to use the default WordPress classes since we use Twig?

**QualifyingElement**
Isn't there a reset / normalize stylesheet that does comply to the Sass styleguide we can use? If we are going to use bower we should include a normalize / reset stylesheet as a dependancy and I don't think we should lint files from dependencies, so that would also solve this issue.

**Comment by haroldangenent** [comment]
**SelectorFormat**
I thought WordPress actually was using lowercases and hyphens. Could you give us an example?

**QualifyingElement**
Normalize is a library, therefore this does not have to be validated like the rest of the styles.

**Comment by haroldangenent** [comment]
This is a common problem when linting: Sass still compiles, even if the linter fails.

**Comment by sboerrigter** [comment]
@haroldangenent I think Sass should still compile if the linter fails. I think it would be very annoying during development if Sass doesn't compile because of a small code-style mistake.

**Comment by haroldangenent** [comment]
@sboerrigter As discussed earlier today with @luukdv, it would be better to stop compilation after bad-code issues (either syntax or code style). This would allow developers to easily oversee problems and commit bad code.

**Comment by sboerrigter** [comment]
I'm still not sure about this, but let's try it.

**Comment by haroldangenent** [comment]
Autoprefixer: See #521.

**Comment by haroldangenent** [comment]
I've just fixed the following SCSS lint errors:
- Vendor prefixes (using Autoprefixer now)
- Id Selectors
- Color literals

SelectorFormat was fixed by @luukdv by editing the SCSS Yaml file. The only problem left is `QualifyingElement` in `_forms.scss`. Could someone take a look at that one?

Please thoroughly review the rest of my adjustments.

**Comment by gewoonderrik** [comment]
@haroldangenent QualifyingElement has been fixed, but as you suggested on Slack there might be a better solution for `_forms.scss`.

**Comment by haroldangenent** [comment]
@gewoonderrik I'll create a separate issue for that.

**Comment by gewoonderrik** [comment]
@haroldangenent All the Linter errors have been resolved now and we're discussing the issues with `_forms.scss` in another topic I think it's allright if we close this issue. How do you feel about this?

**Comment by haroldangenent** [comment]
@gewoonderrik Okay, I'll unassign this. Will be auto-closed when the merge is done.
